### ISSUE TITLE
Interim IX-F JSON importer

### ISSUE THREAD

**Issue by arnoldnipper** [peeringdb/peeringdb#585]
This interim IX-F JSON importer should be in place until the data ownership task-force comes up with a policy. This issue also relates to #518 and #540

There is a tick box "Allow IXP Update" which governs the behaviour of the import. By default, this is unticked (set to "no"). If "Allow IXP Update" is ticked, an entry may be created, changed and removed. The entry then is completely governed by the settings in the IX-F JSON import. 

`allow_ixp_update`: **no**
 * If a network has an IXP entry with differing (asn, ipaddr4, ipaddr6),  **an email is sent** to the `ix` operator, the `net` owner and the Admin committee. This also covers the case that the network does not have an entry in the IXP member list at all
 * If a network has an IXP entry with any other differing information (speed, route server peer), this information is **not changed**
 * If a network does not have an entry for the IXP, **nothing is done**

`allow_ixp_update`: **yes**
 * If a network has an IXP entry with any differing information, the entry is updated (IPv4, IPv6, speed, route server peer)
 * If a network does not have an entry for the IXP, one is added if there is not a conflicting record already. If there is a conflicting record an email is sent to the `ix` operator, the `net` owners and the Admin committee
 * If a network does not have an entry in the IXP member list, the network IXP entry is removed

**Comment by arnoldnipper** [comment]
@nbakker, @fkorsback, @durkovic, @peeingdb/pc and @peeringdb/ac  does that sound like a good compromise?

**Comment by durkovic** [comment]
Sure, that will re-enable the importer for allow_ixp_update:yes users and avoid automatic/unwanted deletions for allow_ixp_update:no users, so hopefully both camps will be happy for now.

Later, in the data ownership TF, I'd suggest to get some more info about data quality, namely:
- how many records are present in IX-F exports but missing in PDB for various reasons
   - network not listing IXP presence
   - network not registered in PDB #391
- how many records have other differences (speed, route server peer)

**Comment by job** [comment]
The (fairly complete!) description of the things that happen when
`allow_ixp_update` is set to true or false is not yet complete, and
should be 100% complete before we can proceed with an implementation.

I think we should collect all permutations possible and describe for
each situation what we expect to happen.

A network may have two connections listed in PDB, while only one exists
in IX-F JSON. The reverse could also exist. When multiple connections
exist, the software should find which connections match, and then figure
out how to procede with 'the rest'.

In describing all these permutations, the format probably should be:
given these circumstances, 'no' will accomplish A, and 'yes' accomplish
'B', rather than use the yes/no key as the primary key to describe what
happens.

To fully resolve this, we need to make sure that we accounted for each
and every possible situation. We can't leave it up to the developer to
try to figure it out, or to some arbitrary result of said code. (Note -
I am not critiquing any developer, just emphasizing that we need to make
sure that we provide them with adequate information to resolve a fairly
complex situation).

I hope the above makes sense.

In any regard, I like the constructive thinking, approaching this
problem thoughtfully (as is happening right now), will surely lead to a
good solution.

**Comment by durkovic** [comment]
@job The purpose of this ticket is to find an emergency interim fix in order to re-enable IX-F importer until the final solution is agreed, designed & implemented.

If you look at https://github.com/peeringdb/peeringdb/blob/master/peeringdb_server/ixf.py
the offending code is in process_deletions() which removes entries not matching (asn,ipaddr4,ipaddr6)

If the above problem is patched by an emergency fix, we'll have enough time to discuss everything in detailm but currently we're lacking vital functionality.

**Comment by job** [comment]
I don’t think we need an emergency interim fix.

Let’s just put in the work to make it decent.

I don’t think it is unreasonable to document how the importer works, people
need that documentation anyhow. The same documentation can be used as
specification on for the developer.

**Comment by durkovic** [comment]
Well, IX-F importer is not working since Aug 30-th and all networks relying on it didn't get updates since then. Automatic updates were publicly announced and are not working for 1.5 months now, so
I do believe we need an emergency fix asap.

The situation with multiple connections you raised works properly in present code, since match is performed on each (asn,ipaddr4,ipaddr6) separately.

Documenting the importer in detail is for sure reasonable, but given a new task force was just created, it might be good to discuss it first and only based on agreement draft the new spec.

**Comment by job** [comment]
That order of work doesn’t work for me.

What do others think?

**Comment by arnoldnipper** [comment]
> The (fairly complete!) description of the things that happen when `allow_ixp_update` is set to true or false is not yet complete and should be 100% complete before we can proceed with an implementation.

Imho my algorithm covers all cases, even with multiple connections to an `ix` as we are always talking about the triple (asn, ipv4,ipv6).

Fully agree that we do not need an emergency fix. But we also did not need an emergency fix to disable the importer ;-)

**Comment by durkovic** [comment]
> Fully agree that we do not need an emergency fix. But we also did not need an emergency fix to disable the importer ;-)

Fair enough :-)
But since disabling of the importer created problems for others, let's try to fix those as well.

**Comment by arnoldnipper** [comment]
@peeringdb/pc could I please have supporters. Tnx!

**Comment by fkorsback** [comment]
The only thing i care about is that when a network set "allow_ixp_update: no" on their profile, they will not get any data automatically deleted in any field anywhere on their profile. 

The only one that should be able to delete a record from an opted-out network is the AC if the automatic script finds out that there is discrepancies between the user-submitted data and what is coming in from the IXF importer. I do understand that other network relies on the IXF importer so i think it's important that we get the function activated. To me it seems like Arnold suggestion is a fair middle ground that should keep both camps happy. 

Then we let the DTF group figure out exactly how these functions will work in the future and how we treat data ownership in the general sense.

**Comment by job** [comment]
Frederik, just as a note: what I’m talking about doesn’t have anything to
do with the DTF.

I’m saying we need a complete specification of what the tool does in what
circumstances, so that users (and volunteers) can understand what the tool
does and how to debug things. Without a complete specification the
developer will have a hard time implementing this.

**Comment by durkovic** [comment]
@job the original spec for IX-F importer is here: 
https://docs.peeringdb.com/ix-f-json-import-rules/

This ticket just replaces **"entry is removed"** with **"an email is sent"**

A that's the whole point of this ticket - avoid removal of entries and generate notification emails instead.

**Comment by arnoldnipper** [comment]
> @job the original spec for IX-F importer is here:
> https://docs.peeringdb.com/ix-f-json-import-rules/
> 
> This ticket just replaces **"entry is removed"** with **"an email is sent"**
> 
> A that's the whole point of this ticket - avoid removal of entries and generate notification emails instead.

Exactly ... the tool is in place since 11/2017 iirc. And as far as I can see it was working bugless so far (seen as tool). 

Two modifications are needed.

 * Notify involved parties in case an entry is marked as "to be deleted" AND "Allow IXP Update " is set to "No"

 * Notify involved parties in case an entry needs to be created AND "Allow IXP Update is set to "Yes" AND IPv4/IPv6 address still is in use

**Comment by eloos** [comment]
+1

**Comment by arnoldnipper** [comment]
To summarize 

`allow_ixp_update`: **no**
  * If a network has an IXP entry with differing (asn, ipaddr4, ipaddr6), **create a DeskPRO ticket**, emailing the network and the ix. (1)

  * If a network has an IXP entry with any other differing information (speed, route server peer), this information is **not changed**

  * If a network does not have an entry for the IXP, **nothing is done**


`allow_ixp_update`: **yes**
  * If a network has an IXP entry with any differing information, the entry is updated (IPv4, IPv6, speed, route server peer) (1)

  * If a network does not have an entry for the IXP,
    * if there is not a conflicting record, a new entry is added.
    * If there is a conflicting record, create a DeskPRO ticket, emailing the networks and the ix.

(1) This also covers the case that the network does not have an entry in the IXP member list at all. In the case of "allow_ixp_update: yes" the update then actually means removal. This prevents involved parties from being flooded by tickets

A DeskPRO ticket will be opened once only and thereby creates a lock. Only if the lock is removed a new ticket can be created. 

@peeringdb/pc please make up your mind

**Comment by funkestefan** [comment]
I like the proposal. 
I would like to see allow_ixp_update as an option for each IXP my network is present at.

**Comment by grizz** [comment]
> I like the proposal.

I do as well, I agree with @job that it might be best to do the specs in a flow chart type style instead of `allow_update` being yes or no. I believe I was the one who wrote the original specs, and there are still ambiguities for things like what to do with multiple IX connections. @arnoldnipper I have some time this month, happy to help go through that with you if you'd like and hammer out better specs.

> I would like to see allow_ixp_update as an option for each IXP my network is present at.

@funkestefan I don't think I agree with that, but either way, it should be a separate ticket with its own discussion, happy to argue further there. :)

**Comment by durkovic** [comment]
I don't think there's any problem with multiple IX connections. Each IX connection must have different IP address, so it's handled separately, one by one.

**Comment by grizz** [comment]
There may not be, point being, specs could use a little more work before implementation or we stand the chance of running into issues down the road.

**Comment by mcmanuss8** [comment]
> To summarize
...
> @peeringdb/pc please make up your mind

+1

**Comment by arnoldnipper** [comment]
@koalafil , this issue hast support from @grizz @eloos @mcmanuss8 @arnoldnipper Please promote to the next stage. Happy to work with @grizz to a proper spec for our vendor.

**Comment by job** [comment]
> I don't think there's any problem with multiple IX connections. Each IX connection must have different IP address, so it's handled separately, one by one.

I foresee issues with multiple connections if we don't specify the workflow:

Imagine if you have connection A & B, both have an IPv4 and IPv6 address - however the IPv4 addresses have been swapped (compared to IX-F), the IPv6 aren't.

**Comment by durkovic** [comment]
> Imagine if you have connection A & B, both have an IPv4 and IPv6 address - however the IPv4 addresses have been swapped (compared to IX-F), the IPv6 aren't.

If the importer checks for exact match on 3-tuples (asn, ipaddr4, ipaddr6), then both records will yield a mismatch and 2 tickets will be created.

**Comment by vegu** [comment]
@arnoldnipper i am trying to estimate the scope for this and have some follow up questions:

For the notifications, am i assuming correctly that you would want one ticket / email sent per ix / network that contains all member issues. Not one ticket per issue?

>A DeskPRO ticket will be opened once only and thereby creates a lock. Only if the lock is removed a new ticket can be created.

Am i understanding this correctly that you want the status of the deskpro ticket inside the deskpro environment to serve as a lock (open vs closed) on whether more notifications are sent. For both for the actual deskpro ticket as well as the notifications to the ix and network? So, as long as the deskpro ticket is open the ix and network wont be notified again either?

If not, what are the specifications for preventing notification spam to ix and network.

**Comment by arnoldnipper** [comment]
Exactly, @vegu. One ticket/email sent per netixlan in question to be more precisely. And as long as the lock is set, no new ticket will be created. Ideally, the lock is removed by AC when the issue is resolved. 

Does that answer all your questions?

**Comment by vegu** [comment]
@arnoldnipper thanks

- emails to exchanges go to tech email ? Do we fall back to other email if it isn't set?
- emails to network goes to pocs with role `Technical` ? (maybe `Maintenance` too?)

Also finally just an idea i'd like to bounce of the wall, feel free to disregard of course

An email per netixlan has the potential to be a lot of emails to a network / exchange - especially right after we roll the feature out.

I wonder if perhaps a mixed approach is more suitable were we do one dekspro ticket per netixlan, but one email per network / ix that contains all the netixlan issues. 

The deskpro ticket lock would still work for both, where only netixlan issues that did not have a deskpro ticket open would be included in the email to the network and or exchange.

**Comment by arnoldnipper** [comment]
> * emails to exchanges go to tech email ? Do we fall back to other email if it isn't set?

`tech_email` || `policy_email`

> * emails to network goes to pocs with role `Technical` ? (maybe `Maintenance` too?)

`Policy` || `Technical` || `NOC` || `Maintenance` ... esp. the larger networks reported that `NOC` typically doesn't understand those emails. Those emails should go to the peering managers.

Regarding the mixed approach, I still tend to have an email per netixlan. Esp. when looking at how the connections are distributed.

 23193 1
   3303 2
    138 3
    138 4
     10 5
     13 6
      2 7
      1 12

Per netixlan would also allow having the `ipaddr{4,6}` information in the subject. Furthermore, the reason for mismatch may differ. Hence a single ticket is way cleaner as it is always clear what issue we are talking about.

Does that make sense?

**Comment by arnoldnipper** [comment]
> * do we want a opt out/in setting for these notification emails?

We want to have a robust procedure. Giving the participants an opt-in/out feature undermines robustness. Hence, the answer is **no**  :)

**Comment by arnoldnipper** [comment]
@Shawna, any reason you pushed this issue back? Please let us know what is still unclear?

**Comment by ccaputo** [comment]
@Shawna and @vegu , if #627 gets approved by the @peeringdb/pc , it may make sense to develop #585 and #627 at the same time since the proposed netixlan_ixp and netixlan_override tables can facilitate the resolution process, by leaving a network's netixlan entry untouched while providing an updated view of the IXP perspective.

**Comment by koalafil** [comment]
Closing in leue of https://github.com/peeringdb/peeringdb/issues/676.
